### DISCUSSION TITLE
Is there a way to pass extra information down the AssistantRuntimeProvider?

### DISCUSSION THREAD

**Discussion by osseonews** [assistant-ui/assistant-ui#1113] [Q&A] [answered]
We have additional information from our API  in our runtime ModelAdapter that we need to pass to the Assistant. right now, we are appending it to the content text in the Runtime, but this causes us many issues. We are trying to figure out a way to pass this information and then be able to access it with useThread or one of the other hooks. Is this possible? Any "hacks" to get this to work?

**Comment by Yonom** [comment, marked as answer]
Is this information per-message or per-thread?

For per-message metadata, your ChatModelAdapter can pass information by returning `{ metadata: { custom: { yourKey: yourValue } }` next to the content. This can be later retrieved via `useMessage(m => m.metadata.custom)`

For per-thread metadata, you might want to simply add a new react context provider with your custom data:
https://react.dev/learn/passing-data-deeply-with-context


Next to AssistantRuntimeProvider you have:
```tsx
const MyDataContext = createContext();

const Providers = () => {
  <AssistantRuntimeProvider>
    <MyDataContext.Provider data={...}
```

https://react.dev/learn/passing-data-deeply-with-context

**Reply by osseonews** [reply]
Thank you so much! The metadata was exactly what we were looking for. I didn't see it in the type, though. I guess looking in wrong place. BTW, we did try our own context provider, but metadata is much less complicated and works perfectly for us.

**Reply by osseonews** [reply]
Sorry, once quick follow up. How could we add the metadata to the copy function at the bottom of the messages in the toolbar? Now the copy only copies the text content, but not the metadata? I couldn't find any documentation on what is sent to the copy or how to adjust it? Thanks.

**Reply by Yonom** [reply]
My suggestion would be to not use AssistantActionBar.Copy / ActionBarPrimitive.Copy but rather make your own button that adds the metadata to the copied message:

```ts
const MyCustomCopy = () => {
  const messageRuntime = useMessageRuntime();
  const setIsCopied = useMessageUtils((s) => s.setIsCopied);

  const handleCopy = () => {
    const metadataToCopy = messageRuntime.getState().metadata.custom.myKey;
    const textToCopy = messageRuntime.unstable_getCopyText();
    const valueToCopy = textToCopy + metadataToCopy;

    navigator.clipboard.writeText(valueToCopy).then(() => {
      setIsCopied(true);
      setTimeout(() => setIsCopied(false), 3000);
    });
  };

  return <button onClick={handleCopy} />

  // alternatively if you want the styling of the default button
  return <AssistantActionBar.Copy onClick={(e) => {
    e.preventDefault(); // disable the default copy behavior
    handleCopy();
  }) />;
}
```

**Reply by Yonom** [reply]
In the future, there will be a way to pass a custom "CopyAdapter" (name TBD) that can alter the result of `unstable_getCopyText` so you don't have to do this :)

**Reply by osseonews** [reply]
Thanks that worked! And thanks again for the amazing library and support. BTW, I noticed you were using Trieve. Recommend looking into Orama. It's easier and I believe better (no relationship, just a customer, and that's what we use with assistant-ui). We build a very rudimentary integration with Orama, as we are not that experienced with Assistant-UI yet, but imagine the Trieve package can be adjusted a bit to match the Orama API and you'd have another integration.

**Reply by Yonom** [reply]
Thanks a lot for the feedback!
